compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -353,39 +342,80 @@ LogicalResult splitLogicalObjectFifos(
+  SmallVector<OpFoldResult> staticL2AsTargetOffsets =
       l3ToL2DmaOp.getTargetMixedOffsets();
-  SmallVector<OpFoldResult, 4> staticL2AsTargetSizes =
+  SmallVector<OpFoldResult> staticL2AsTargetSizes =
       l3ToL2DmaOp.getTargetMixedSizes();
-  SmallVector<OpFoldResult, 4> staticL3AsSourceOffsets =
+  SmallVector<OpFoldResult> staticL3AsSourceOffsets =
       l3ToL2DmaOp.getSourceMixedOffsets();
-  SmallVector<OpFoldResult, 4> staticL3AsSourceSizes =
+  SmallVector<OpFoldResult> staticL3AsSourceSizes =
       l3ToL2DmaOp.getSourceMixedSizes();
+
+  LogicalObjectFifoFromMemrefOp l2TargetObjectFifo =
+      l3ToL2DmaOp.getTargetObjectFifo();
+  ArrayRef<int64_t> l2TargetShape =
+      l2TargetObjectFifo.getMemrefType().getShape();
+  if (l2TargetShape.size() != staticL2AsTargetSizes.size()) {
+    LLVM_DEBUG(llvm::dbgs() << "L2 target size should be the same");
+    return failure();
+  }
+
+  // Check if the L3->L2 dma is transposed on the target side.
+  bool dmaTransposeOnSource = true;
+  for (auto [s1, s2] : llvm::zip_equal(l2TargetShape, staticL2AsTargetSizes)) {
+    if (s1 != getConstantIntValue(s2)) {
+      dmaTransposeOnSource = false;
+    }
+  }
+
   OpFoldResult zeroVal = getAsIndexOpFoldResult(context, 0);
   OpFoldResult oneVal = getAsIndexOpFoldResult(context, 1);
-  // Update split dimensions' offset/size for L2 as target and L3 as source. We
-  // can afford to do this here because it's going to be the same for all L3->L2
-  // splits. Here we are setting offset = 0 and size = 1.
-  for (size_t dim : splitDimsForL2) {
-    staticL2AsTargetOffsets[dim] = zeroVal;
-    staticL2AsTargetSizes[dim] = oneVal;
-    staticL3AsSourceOffsets[dim] = zeroVal;
-    staticL3AsSourceSizes[dim] = oneVal;
+
+  if (dmaTransposeOnSource) {
+    // Update split dimensions' offset/size for L2 as target and L3 as source.
+    // We can afford to do this here because it's going to be the same for all
+    // L3->L2 splits. Here we are setting offset = 0 and size = 1.
+    for (size_t dim : splitDimsForL2) {
+      staticL2AsTargetOffsets[dim] = zeroVal;
+      staticL2AsTargetSizes[dim] = oneVal;
+      staticL3AsSourceOffsets[dim] = zeroVal;
+      staticL3AsSourceSizes[dim] = oneVal;
+    }
+  } else {
+    // The L2 target side has transposed dimensions, while the L3 source side
+    // data are continuous and doesn't have "nonSplitDim".
+    // Hardcoded the transposed dimensions for now.
+    const SmallVector<size_t> transposeDim = {0, 2, 1, 3};
+    for (size_t dim : splitDimsForL2) {
+      staticL2AsTargetOffsets[transposeDim[dim]] = zeroVal;
+      staticL2AsTargetSizes[transposeDim[dim]] = oneVal;
+    }
+    // Modify the L3 source sizes to match the new L2 target sizes.
+    for (auto &&[splitDim, nonSplitdim] :
+         llvm::zip_equal(splitDimsForL2, nonSplitDimsForL2)) {
+      staticL3AsSourceSizes[splitDim] =
+          staticL2AsTargetSizes[transposeDim[nonSplitdim]];
+    }
```

Comment:
Can we try to fuse these two into a single loop with :- 
```
 for (auto &&[splitDim, nonSplitdim] : llvm::zip_equal(splitDimsForL2, nonSplitDimsForL2))
```
Or is there a dependency which gets invalidated ?

Reply:
Done.